mountpoint-s3/src/data_cache/disk_data_cache.rs

```diff
@@ -0,0 +1,333 @@
+    /// However, this check is not guaranteed and it shouldn't be assumed that the data within the block is not corrupt.
+    fn new(cache_key: CacheKey, block_idx: BlockIndex, bytes: ChecksummedBytes) -> DataCacheResult<Self> {
+        let (s3_key, etag) = (cache_key.s3_key, cache_key.etag.into_inner());
+        let (data, checksum) = bytes.into_inner().map_err(|_e| DataCacheError::InvalidBlockContent)?;
+        let checksum = checksum.value();
+        Ok(DataBlock {
+            block_idx,
+            checksum,
+            data,
+            etag,
+            s3_key,
+        })
+    }
+
+    /// Extract the block data, checking that fields such as S3 key, etc. match what we expect.
+    ///
+    /// Comparing these fields helps ensure we have not corrupted or swapped block data on disk.
+    fn data(&self, cache_key: &CacheKey, block_idx: BlockIndex) -> DataCacheResult<ChecksummedBytes> {
+        let (s3_key, etag) = (cache_key.s3_key.as_str(), &cache_key.etag);
+
+        let s3_key_match = s3_key == self.s3_key;
+        let etag_match = etag.as_str() == self.etag;
+        let block_idx_match = block_idx == self.block_idx;
+        if s3_key_match && etag_match && block_idx_match {
+            let bytes = ChecksummedBytes::new(self.data.clone(), Crc32c::new(self.checksum));
+            Ok(bytes)
+        } else {
+            error!(
+                s3_key_match,
+                etag_match, block_idx_match, "block data did not match expected values",
+            );
+            Err(DataCacheError::InvalidBlockContent)
+        }
+    }
+}
+
+impl DiskDataCache {
+    /// Create a new instance of an [DiskDataCache] with the specified `block_size`.
+    pub fn new(cache_directory: PathBuf, block_size: u64) -> Self {
+        DiskDataCache {
+            block_size,
+            cache_directory,
+        }
+    }
+
+    /// Get the relative path for the given block.
+    fn get_path_for_key(&self, cache_key: &CacheKey) -> PathBuf {
+        let mut path = self.cache_directory.join(CACHE_VERSION);
+
+        // An S3 key may be up to 1024 UTF-8 bytes long, which exceeds the maximum UNIX file name length.
+        // Instead, we hash the key.
+        // The risk of collisions is mitigated as we will ignore blocks read that contain the wrong S3 key, etc..
+        let encoded_s3_key = hex::encode(Sha256::digest(cache_key.s3_key.as_bytes()));
+        // TODO: Split directory into subdirectories.
+        //       Take the first few chars of hash to avoid hitting any FS-specific maximum number of directory entries.
+        path.push(encoded_s3_key);
+        path.push(cache_key.etag.as_str());
+        path
+    }
+
+    /// Get path for the given block.
+    fn get_path_for_block(&self, cache_key: &CacheKey, block_idx: BlockIndex) -> PathBuf {
+        let mut path = self.get_path_for_key(cache_key);
+        path.push(format!("{}.block", block_idx));
+        path
+    }
```

Comment:
Did you decide what you want to do here w.r.t. including the etag and block index in the hash or not?

Reply:
I'm going to leave it as is for this PR. I think its worth leaving the block index out of the hash, so that we can wipe out all the blocks for a given S3 & ETag pair.

We can have more discussion in the PR that moves in that direction.

Reply:
Opened #594.